tgui/packages/tgui/interfaces/GeneTek.js

```diff
@@ -0,0 +1,1316 @@
+              act("unlock", { code: "UNLOCK" });
+            }}>
+            Use Auto-Decryptor
+          </Button>
+        </Box>
+        <Box textAlign="right" mt={1}>
+          <Button
+            mr={1}
+            icon="check"
+            color="good"
+            tooltip={unlockCode.length !== unlock.length
+              ? "Code is the wrong length."
+              : unlockCode.split("").some(c => unlock.chars.indexOf(c) === -1)
+                ? "Invalid character in code." : ""}
+            disabled={unlockCode.length !== unlock.length
+              || unlockCode.split("").some(c => unlock.chars.indexOf(c) === -1)}
+            onClick={() => {
+              setUnlockCode("");
+              act("unlock", { code: unlockCode });
+            }}>
+            Attempt Decryption
+          </Button>
+          <Button
+            icon="times"
+            color="bad"
+            onClick={() => {
+              setUnlockCode("");
+              act("unlock", { code: null });
+            }}>
+            Cancel
+          </Button>
+        </Box>
+      </Box>
+    </Modal>
+  );
+};
+
+const letterColor = {
+  "?": "grey",
+  "A": "red",
+  "T": "blue",
+  "C": "yellow",
+  "G": "green",
+};
+
+const typeColor = {
+  "": "good",
+  "X": "grey",
+  "1": "good",
+  "2": "olive",
+  "3": "average",
+  "4": "orange",
+  "5": "bad",
+};
+
+const DNASequence = (props, context) => {
+  const { act } = useBackend(context);
+  const {
+    gene,
+    isPotential,
+  } = props;
+
+  const sequence = gene.dna;
+  let allGood = true;
+
+  const blocks = [];
+  for (let i = 0; i < sequence.length; i++) {
+    if (i % 4 === 0) {
+      blocks.push([]);
+    }
+
+    blocks[blocks.length - 1].push(sequence[i]);
+
+    if (sequence[i][2]) {
+      allGood = false;
+    }
+  }
+
+  const advancePair = i => {
+    if (isPotential) {
+      act("advancepair", {
+        ref: gene.ref,
+        pair: i,
+      });
+    }
+  };
+
+  return blocks.map((block, i) => (
+    <table key={i} style={{
+      display: "inline-table",
+      "margin-top": "1em",
+      "margin-left": i % 4 === 0 ? "0" : "0.25em",
+      "margin-right": i % 4 === 3 ? "0" : "0.25em",
+    }}>
+      <tr>
+        {block.map((pair, j) => (
+          <td key={j}>
+            <Nucleotide
+              letter={pair[0]}
+              type={pair[2]}
+              mark={pair[3]}
+              useLetterColor={allGood}
+              onClick={() => advancePair(i * 4 + j + 1)} />
+          </td>
+        ))}
+      </tr>
+      <tr>
+        {block.map((pair, j) => (
+          <td key={j} style={{ "text-align": "center" }}>
+            {allGood ? "|" : pair[3] === "locked" ? (
+              <Icon
+                name="lock"
+                color="average"
+                onClick={() => advancePair(i * 4 + j + 1)} />
+            ) : (
+              <Icon name={
+                pair[2] === "" ? "check"
+                  : pair[2] === "5" ? "times"
```

Comment:
What is the significance of `"5"`?

Reply:
The number ranges from 1 to 5, with 1 as green, 3 as yellow, and 5 as red.

Reply:
Rephrasing: this info should be added to the code.